flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommittableCollector.java

```diff
@@ -0,0 +1,235 @@
+    private final NavigableMap<Long, CheckpointCommittablesImpl<CommT>> checkpointCommittables;
+    private final int subtaskId;
+    private final int numberOfSubtasks;
+
+    CommittableCollector(int subtaskId, int numberOfSubtasks) {
+        this.subtaskId = subtaskId;
+        this.numberOfSubtasks = numberOfSubtasks;
+        this.checkpointCommittables = new TreeMap<>();
+    }
+
+    /** For deserialization. */
+    CommittableCollector(Map<Long, CheckpointCommittablesImpl<CommT>> checkpointCommittables) {
+        this(checkpointCommittables, 0, 1);
+    }
+
+    /** For deep-copy. */
+    CommittableCollector(
+            Map<Long, CheckpointCommittablesImpl<CommT>> checkpointCommittables,
+            int subtaskId,
+            int numberOfSubtasks) {
+        this.checkpointCommittables = new TreeMap<>(checkNotNull(checkpointCommittables));
+        this.subtaskId = subtaskId;
+        this.numberOfSubtasks = numberOfSubtasks;
+    }
+
+    /**
+     * Creates a {@link CommittableCollector} based on the current runtime information. This method
+     * should be used for to instantiate a collector for all Sink V2.
+     *
+     * @param context holding runtime of information
+     * @param <CommT> type of the committable
+     * @return {@link CommittableCollector}
+     */
+    public static <CommT> CommittableCollector<CommT> of(RuntimeContext context) {
+        return new CommittableCollector<>(
+                context.getIndexOfThisSubtask(), context.getNumberOfParallelSubtasks());
+    }
+
+    /**
+     * Creates a {@link CommittableCollector} for a list of committables. This method is mainly used
+     * to create a collector from the state of Sink V1.
+     *
+     * @param r list of committables
+     * @param <CommT> type of committables
+     * @return {@link CommittableCollector}
+     */
+    static <CommT> CommittableCollector<CommT> ofLegacy(List<CommT> r) {
+        CommittableCollector<CommT> committableCollector = new CommittableCollector<>(0, 1);
+        // add a checkpoint with the lowest checkpoint id, this will be merged into the next
+        // checkpoint data, subtask id is arbitrary
+        CommittableSummary<CommT> summary =
+                new CommittableSummary<>(
+                        0, 1, InitContext.INITIAL_CHECKPOINT_ID, r.size(), r.size(), 0);
+        committableCollector.addSummary(summary);
+        SubtaskCommittables<CommT> subtask =
+                committableCollector
+                        .getCheckpointCommittables(summary)
+                        .getSubtaskCommittables(summary.getSubtaskId());
+        r.forEach(subtask::add);
+        return committableCollector;
+    }
+
+    /**
+     * Adds a {@link CommittableMessage} to the collector to hold it until emission.
+     *
+     * @param message either {@link CommittableSummary} or {@link CommittableWithLineage}
+     */
+    public void addMessage(CommittableMessage<CommT> message) {
```

Comment:
Nit

Reply:
I'd prefer to keep it short.

Reply:
sure, it is not a big deal. The long name will improve the code readability at the place where the method is called, since code like `addMessage(message)` does not contain enough info. But once the developer get familiar with the code, it is not an issue.